Convert testclusters to use distro download plugin

rjernst opened the pull request:
Test clusters currently has its own set of logic for dealing with
finding different versions of Elasticsearch, downloading them, and
extracting them. This commit converts testclusters to use the
DistributionDownloadPlugin.

rjernst committed:
Convert testclusters to use distro download plugin

Test clusters currently has its own set of logic for dealing with
finding different versions of Elasticsearch, downloading them, and
extracting them. This commit converts testclusters to use the
DistributionDownloadPlugin.

elasticmachine commented:
Pinging @elastic/es-core-infra

rjernst committed:
Merge branch 'master' into distro_tests13

alpar-t reviewed (changes requested):
Looks great! 

I have one major concern as it seems we are loosing an important optimization with this change.

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchCluster.java:
```diff
@@ -60,22 +60,23 @@
     private final NamedDomainObjectContainer<ElasticsearchNode> nodes;
     private final File workingDirBase;
-    private final File artifactsExtractDir;
+    private final Function<Integer, ElasticsearchDistribution> distributionFactory;
```
I added a comment here previously which seems to have vanished. 

It seems that we are generating a distribution per node, defeating the purpose of `syncWithLinks`.
This optimization is important both in terms of disk space and time the build takes and I would prefer we don't merge this without it.  

It might make more sense to have this optimization in the distribution downloaded plugin trough and have test cluusters use the extracted distro dirs directly.

rjernst replied:
The distribution download plugin already optimizes by only doing a single extraction for a given set of distribution properties. While we create a unique ElasticsearchDistribution object per node here, the underlying extracted dir will just exist once, in the root project.

atorok replied:
Thanks, got it, we only use the index because we need unique names for the domain object container.  
I wonder if we should be concerned about name clashes here ? In theory other uses of the distribution download plugin could use the same name as testclusters even trough unlikely.

mark-vieira replied:
Clashes are possible but will result in an obvious error. I don't think we need to be too concerned about it.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java:
```diff
@@ -94,25 +82,23 @@ public void apply(Project project) {
+
+        project.getPlugins().apply(DistributionDownloadPlugin.class);
+        @SuppressWarnings("unchecked")
+        NamedDomainObjectContainer<ElasticsearchDistribution> distros =
```
You can. You have to extend `AbstractNamedDomainObjectContainer` which is technically internal but it's not an uncommon thing to do. The other option is to just create a default container, and delegate to it from your custom one. This is cumbersome in Java but could be done easily w/o boilderplate if the implementation was Groovy using something like `@Delegate`.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java:
```diff
@@ -974,19 +994,19 @@ private Path getExtractedDistributionDir() {
-        return files;
+    private Iterable<File> getDistributionClasspath() {
+        FileTree jarFiles = distribution.getExtracted().getFileTree(project).matching(filter -> filter.include("**/*.jar"));
+        return () -> {
```
This is very clever 👍. Thing is, it doesn't actually do what we want, lol. So my intuition was incorrect and returning an `Iterable` doesn't actually defer anything at task graph calculation time. That's because this iterable could be some other `Buildable` type, Gradle has to iterate this collection and add each entry as input. Due to type erasure we have no idea at runtime what this is a collection of, so we have to rip through it.

The interesting thing here though is that our assumption that calling `getSingleFile()` would cause resolution is actually incorrect. This doesn't actually trigger any work. It just returns a path to the empty directory as we are just resolving the configuration, but not it's _artifacts_. So all this effort to defer that was in vain 🤦‍♂ .

With that knowledge in hand we might be able to simplify some of this logic now that we knowing calling `getSingleFile()` on a configuration which includes a single directory artifact is fine.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java:
```diff
@@ -94,25 +82,23 @@ public void apply(Project project) {
 
     private NamedDomainObjectContainer<ElasticsearchCluster> createTestClustersContainerExtension(Project project) {
+
+        project.getPlugins().apply(DistributionDownloadPlugin.class);
```
I'd move this up to the first line of the plugin `apply()` method. It makes it very obvious that this other plugin is a "dependency" of this plugin.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchCluster.java:
```diff
@@ -338,9 +339,9 @@ public boolean isProcessAlive() {
     }
 
-    void eachVersionedDistribution(BiConsumer<String, Distribution> consumer) {
+    void eachVersionedDistribution(BiConsumer<Version, TestDistribution> consumer) {
```
You're right. This can now be removed.

mark-vieira reviewed (approved):
Few comments, otherwise this looks great! Running tests that only use the integ-test distro are much faster now and overall this PR remove more code than it added which is always a step in the right direction.

rjernst committed:
address feedback, move default distro and version into node

rjernst commented:
@mark-vieira @atorok I think I have addressed all your feedback. One small change I had to make was to move how the test distribution is defaulted. Before this happened inside RestIntegTestTask. However, that required using an afterEvaluate to set it for qa tests, which use the oss distro (and have a sysprop to change it for all qa tests). Using the integ test zip by default is now done directly within ElasticsearchNode (see https://github.com/elastic/elasticsearch/pull/44253/files#diff-a704b6e9e08a9ccfcc05e2ae59844495R163).

rjernst committed:
use setters for setting defaults

rjernst committed:
solidify naming of test distro enum

rjernst committed:
fix third party audit tests

rjernst commented:
@elasticmachine run elasticsearch-ci/2

rjernst commented:
@elasticmachine run elasticsearch-ci/bwc

rjernst commented:
@elasticmachine run elasticsearch-ci/bwc

rjernst committed:
Merge branch 'master' into distro_tests13

alpar-t reviewed (approved):
LGTM
